File: subprojects/scala/src/main/scala/codechicken/mixin/scala/MixinScalaLanguageSupport.scala
package codechicken.mixin.scala

import java.util.Optional

import codechicken.mixin.api.{MixinCompiler, MixinLanguageSupport}
import codechicken.mixin.scala.MixinScalaLanguageSupport._
import codechicken.mixin.util.{ClassInfo, FieldMixin, MixinInfo}
import org.apache.logging.log4j.LogManager
import org.objectweb.asm.Opcodes._
import org.objectweb.asm.Type
import org.objectweb.asm.Type._
import org.objectweb.asm.tree.{ClassNode, MethodNode}

import scala.collection.mutable.{ListBuffer => MList}
import scala.jdk.CollectionConverters._


/**
 * Default implementation for Scala.
 *
 * Created by covers1624 on 2/11/20.
 */
@MixinLanguageSupport.LanguageName("scala")
class MixinScalaLanguageSupport(mixinCompiler: MixinCompiler) extends MixinLanguageSupport {

    def listFiltered(sig: ScalaSignature): Set[String] = {
        sig.collect[sig.AnnotationInfo](40).filter { a =>
            val value = a.getValue[sig.Literal]("value")
            value.isInstanceOf[sig.EnumLiteral] && mixinCompiler.getMixinBackend.filterMethodAnnotations(a.annType.name, value.asInstanceOf[sig.EnumLiteral].value.full)
        }.map(_.owner.full).toSet
    }

    override def obtainInfo(name: String, cNode: ClassNode): Optional[ClassInfo] = {
        if (cNode == null) {
            return None
        }

        def scalaInfo(cnode: ClassNode, obj: Boolean) =
            ScalaSigReader.ann(cnode).flatMap { ann =>
                val sig = ScalaSigReader.read(ann)
                val name = cnode.name.replace('/', '.')
                (if (obj) sig.findObject(name) else sig.findClass(name))
                    .map(csym => ScalaClassInfo(mixinCompiler, cnode, sig, csym))
            }

        if (cNode.name.endsWith("$")) {
            //find scala object
            val baseName = cNode.name.substring(0, cNode.name.length - 1)
            val baseNode = mixinCompiler.getClassNode(baseName)
            if (baseNode != null) {
                scalaInfo(baseNode, true) match {
                    case Some(info) => return Some(info)
                    case None =>
                }
            }
        }
        scalaInfo(cNode, false)
    }

    def getAndRegisterParentTraits(cnode: ClassNode) = cnode.interfaces.asScala.map(mixinCompiler.getClassInfo).collect {
        case i: ScalaClassInfo if i.isTrait && !i.cSym.isInterface =>
            mixinCompiler.registerTrait(i.getCNode)
    }

    override def buildMixinTrait(cNode: ClassNode): Optional[MixinInfo] = {
        //If we aren't a scala class, we don't do anything.
        val info = mixinCompiler.getClassInfo(cNode) match {
            case info: ScalaClassInfo if info.isTrait => info
            case _ => return None
        }

        val sig = info.sig
        val filtered = listFiltered(sig)

        val parentTraits = getAndRegisterParentTraits(cNode)
        val fields = MList[FieldMixin]()
        val methods = MList[MethodNode]()
        val supers = MList[String]()

        val csym = info.cSym
        for (sym <- sig.collect[sig.MethodSymbol](8)) {
            logger.debug(sym)
            if (sym.isParam || sym.owner != csym) {}
            else if (filtered(sym.full)) {}
            else if (sym.isAccessor) {
                if (!sym.name.trim.endsWith("$eq")) {
                    fields += new FieldMixin(sym.name.trim, getReturnType(sym.jDesc).getDescriptor,
                        if (sym.isPrivate) ACC_PRIVATE else ACC_PUBLIC)
                }
            }
            else if (sym.isMethod) {
                val desc = sym.jDesc
                if (sym.name.startsWith("super$")) {
                    supers += sym.name.substring(6) + desc
                } else if (!sym.isPrivate && !sym.isDeferred && sym.name != "$init$") {
                    //also check if the return type is a scala object
                    val objectDesc = Type.getMethodDescriptor(Type.getObjectType(Type.getReturnType(desc).getInternalName + "$"), Type.getArgumentTypes(desc): _*)
                    methods += (cNode.methods.asScala.find(m => m.name == sym.name && (m.desc == desc || m.desc == objectDesc)) match {
                        case Some(m) => m
                        case None => throw new IllegalArgumentException("Unable to add mixin trait " + cNode.name + ": " +
                            sym.name + desc + " found in scala signature but not in class file. Most likely an obfuscation issue.")
                    })
                }
            }
        }
        Some(new MixinInfo(cNode.name, csym.jParent, parentTraits.toList.asJava, fields.asJava, methods.asJava, supers.asJava))
    }
}

object MixinScalaLanguageSupport {

    import scala.language.implicitConversions

    var logger = LogManager.getLogger()

    //Auto convert scala Option to Optional
    implicit def asJavaOptional[T](op: Option[T]): Optional[T] = op match {
        case Some(e) => Optional.of(e)
        case None => Optional.empty()
    }

    implicit def asScalaOption[T](op: Optional[T]): Option[T] = if (op.isPresent) Some(op.get) else None

}

File: modules/scala/src/main/java/codechicken/mixin/scala/MixinScalaLanguageSupport.java
package codechicken.mixin.scala;

import codechicken.mixin.api.MixinCompiler;
import codechicken.mixin.api.MixinLanguageSupport;
import codechicken.mixin.scala.ScalaSignature.ClassSymbolRef;
import codechicken.mixin.util.ClassInfo;
import codechicken.mixin.util.FieldMixin;
import codechicken.mixin.util.MixinInfo;
import net.covers1624.quack.collection.FastStream;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Nullable;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import static codechicken.mixin.api.MixinLanguageSupport.LanguageName;
import static org.objectweb.asm.Opcodes.ACC_PRIVATE;
import static org.objectweb.asm.Opcodes.ACC_PUBLIC;

/**
 * Created by covers1624 on 19/1/24.
 */
@LanguageName ("scala")
public class MixinScalaLanguageSupport implements MixinLanguageSupport {

    private static final Logger LOGGER = LogManager.getLogger();

    private final MixinCompiler mixinCompiler;

    public MixinScalaLanguageSupport(MixinCompiler mixinCompiler) {
        this.mixinCompiler = mixinCompiler;
    }

    @Override
    public ClassInfo obtainInfo(ClassNode cNode) {
        if (cNode.name.endsWith("$")) {
            String baseName = cNode.name.substring(0, cNode.name.length() - 1);
            ClassNode baseNode = mixinCompiler.getClassNode(baseName);
            if (baseNode != null) {
                ScalaClassInfo info = scalaInfo(baseNode, true);
                if (info != null) {
                    return info;
                }
            }
        }

        return scalaInfo(cNode, false);
    }

    private @Nullable ScalaClassInfo scalaInfo(ClassNode cNode, boolean obj) {
        ScalaSignature sig = ScalaSignature.parse(cNode);
        if (sig == null) return null;

        String name = cNode.name.replace('/', '.');
        ClassSymbolRef cSym = obj ? sig.findObject(name) : sig.findClass(name);
        if (cSym == null) return null;

        return new ScalaClassInfo(mixinCompiler, cNode, sig, cSym);
    }

    @Override
    public MixinInfo buildMixinTrait(ClassNode cNode) {
        if (!(mixinCompiler.getClassInfo(cNode) instanceof ScalaClassInfo info) || !info.isTrait()) return null;

        ScalaSignature sig = info.sig;
        Set<String> filtered = listFiltered(sig);

        List<MixinInfo> parentTraits = getAndRegisterParentTraits(cNode);
        List<FieldMixin> fields = new ArrayList<>();
        List<MethodNode> methods = new ArrayList<>();
        List<String> supers = new ArrayList<>();

        ClassSymbolRef cSym = info.cSym;
        for (ScalaSignature.MethodSymbol sym : sig.<ScalaSignature.MethodSymbol>collect(8)) {
            LOGGER.info(sym);
            if (sym.isParam() || !sym.owner().equals(cSym)) continue;
            if (filtered.contains(sym.full())) continue;

            if (sym.isAccessor()) {
                if (!sym.name().trim().endsWith("$eq")) {
                    fields.add(new FieldMixin(sym.name().trim(), Type.getReturnType(sym.jDesc()).getDescriptor(), sym.isPrivate() ? ACC_PRIVATE : ACC_PUBLIC)); // TODO better
                }
            } else if (sym.isMethod()) {
                String desc = sym.jDesc();
                if (sym.name().startsWith("super$")) {
                    supers.add(sym.name().substring(6) + desc);
                } else if (!sym.isPrivate() && !sym.isDeferred() && !sym.name().equals("$init$")) {
                    // also check if the return type is a scala object
                    String objectDesc = Type.getMethodDescriptor(
                            Type.getObjectType(Type.getReturnType(desc).getInternalName() + "$"),
                            Type.getArgumentTypes(desc)
                    );
                    MethodNode mNode = FastStream.of(cNode.methods)
                            .filter(e -> e.name.equals(sym.name()) && (e.desc.equals(desc) || e.desc.equals(objectDesc)))
                            .firstOrDefault();
                    if (mNode == null) {
                        throw new IllegalArgumentException("Unable to add mixin trait " + cNode.name + ": " + sym.name() + desc + " found in scala signature but not in class file. Most likely an obfuscation issue.");
                    }
                    methods.add(mNode);
                }
            }
        }

        return new MixinInfo(cNode.name, cSym.jParent(), parentTraits, fields, methods, supers);
    }

    private List<MixinInfo> getAndRegisterParentTraits(ClassNode cNode) {
        return FastStream.of(cNode.interfaces)
                .map(mixinCompiler::getClassInfo)
                .filter(e -> e instanceof ScalaClassInfo info && info.isTrait() && !info.cSym.isInterface())
                .map(e -> mixinCompiler.registerTrait(((ScalaClassInfo) e).getCNode()))
                .toList();
    }

    private Set<String> listFiltered(ScalaSignature sig) {
        return FastStream.of(sig.<ScalaSignature.AnnotationInfo>collect(40))
                .filter(e -> {
                    ScalaSignature.Literal value = e.getValue("value");
                    return value instanceof ScalaSignature.EnumLiteral lit
                           && mixinCompiler.getMixinBackend().filterMethodAnnotations(e.annType().name(), lit.value().full());
                })
                .map(e -> e.owner().full())
                .toSet();
    }

}
